File: 3729-unit-conversion-i/3729-unit-conversion-i.cpp
class Solution {
public:
    vector<int> baseUnitConversions(vector<vector<int>>& conversions) {
        const int mod = 1e9 + 7;
        unordered_map<int, list<pair<int, int>>> adj;
        for (auto& conversion : conversions) {
            int f = conversion[0];
            int t = conversion[1];
            int r = conversion[2];
            adj[f].push_back({t, r});
        }
        int n = conversions.size();
        vector<int> ans(n + 1, -1);
        ans[0] = 1;
        queue<int> q;
        q.push (0);
        while (!q.empty()) {
            auto top = q.front();
            q.pop();
            for (auto& neigh : adj[top]) {
                if (ans[neigh.first] == -1) {
                    ans[neigh.first] = (1LL * ans[top] * neigh.second) % mod;
                    q.push (neigh.first);
                }
            }
        }
        return ans;
    }
};

File: 3729-unit-conversion-i/3729-unit-conversion-i.java
class Solution {
    class Edge {
        int t;
        int r;
        Edge(int t, int r) {
            this.t = t;
            this.r = r;
        }
    }
    public int[] baseUnitConversions(int[][] conversions) {
        final int mod = 1_000_000_007;
        HashMap<Integer, List<Edge>> adj = new HashMap<>();
        for (int[] conversion : conversions) {
            int f = conversion[0];
            int t = conversion[1];
            int r = conversion[2];
            adj.computeIfAbsent(f, k -> new ArrayList<>()).add(new Edge(t, r));
        }
        int n = conversions.length;
        int[] ans = new int[n + 1];
        ans[0] = 1;
        
        Queue<Integer> q = new ArrayDeque<>();
        q.offer (0);
        while (!q.isEmpty()) {
            int top = q.poll();
            for (Edge eg : adj.getOrDefault(top, Collections.emptyList())) {
                if (ans[eg.t] == 0) {
                    ans[eg.t] = (int) (((long) ans[top] * eg.r) % mod);
                    q.offer (eg.t);
                }
            }
        }
        return ans;
        
    }
}